File: sdk/java-holons/src/main/java/org/organicprogramming/holons/HolonRPCClient.java
package org.organicprogramming.holons;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.time.Duration;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Holon-RPC JSON-RPC 2.0 client over WebSocket using the {@code holon-rpc}
 * subprotocol.
 */
public final class HolonRPCClient {
    /**
     * Handler for server-initiated calls.
     */
    @FunctionalInterface
    public interface Handler {
        CompletionStage<Map<String, Object>> handle(Map<String, Object> params);
    }

    /**
     * JSON-RPC error response exception.
     */
    public static final class HolonRPCResponseException extends RuntimeException {
        private final int code;
        private final Object data;

        public HolonRPCResponseException(int code, String message, Object data) {
            super("rpc error " + code + ": " + message);
            this.code = code;
            this.data = data;
        }

        public int code() {
            return code;
        }

        public Object data() {
            return data;
        }
    }

    private static final Type MAP_TYPE = new TypeToken<Map<String, Object>>() {
    }.getType();

    private final HttpClient httpClient;
    private final Gson gson;

    private final long heartbeatIntervalMs;
    private final long heartbeatTimeoutMs;
    private final long reconnectMinDelayMs;
    private final long reconnectMaxDelayMs;
    private final double reconnectFactor;
    private final double reconnectJitter;
    private final long connectTimeoutMs;
    private final long requestTimeoutMs;

    private final ScheduledExecutorService scheduler;
    private final Map<String, Handler> handlers = new ConcurrentHashMap<>();
    private final Map<String, CompletableFuture<Map<String, Object>>> pending = new ConcurrentHashMap<>();
    private final AtomicLong nextID = new AtomicLong(0);

    private volatile String endpoint;
    private volatile WebSocket socket;
    private volatile ScheduledFuture<?> heartbeatTask;
    private volatile ScheduledFuture<?> reconnectTask;
    private volatile CompletableFuture<Void> connectedFuture = new CompletableFuture<>();
    private volatile boolean closed = true;
    private volatile int reconnectAttempt = 0;

    public HolonRPCClient() {
        this(15_000, 5_000, 500, 30_000, 2.0, 0.1, 10_000, 10_000);
    }

    public HolonRPCClient(
            long heartbeatIntervalMs,
            long heartbeatTimeoutMs,
            long reconnectMinDelayMs,
            long reconnectMaxDelayMs,
            double reconnectFactor,
            double reconnectJitter,
            long connectTimeoutMs,
            long requestTimeoutMs) {
        this.heartbeatIntervalMs = heartbeatIntervalMs;
        this.heartbeatTimeoutMs = heartbeatTimeoutMs;
        this.reconnectMinDelayMs = reconnectMinDelayMs;
        this.reconnectMaxDelayMs = reconnectMaxDelayMs;
        this.reconnectFactor = reconnectFactor;
        this.reconnectJitter = reconnectJitter;
        this.connectTimeoutMs = connectTimeoutMs;
        this.requestTimeoutMs = requestTimeoutMs;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofMillis(connectTimeoutMs))
                .build();
        this.gson = new Gson();
        this.scheduler = Executors.newSingleThreadScheduledExecutor(runnable -> {
            Thread thread = new Thread(runnable, "holon-rpc-client");
            thread.setDaemon(true);
            return thread;
        });
    }

    public void connect(String url) throws Exception {
        Objects.requireNonNull(url, "url");
        if (url.isBlank()) {
            throw new IllegalArgumentException("url is required");
        }

        close();
        closed = false;
        endpoint = url;
        connectedFuture = new CompletableFuture<>();

        openSocket(true);
        awaitConnected(connectTimeoutMs);
    }

    public void register(String method, Handler handler) {
        Objects.requireNonNull(method, "method");
        Objects.requireNonNull(handler, "handler");
        if (method.isBlank()) {
            throw new IllegalArgumentException("method is required");
        }
        handlers.put(method, handler);
    }

    public Map<String, Object> invoke(String method, Map<String, Object> params) throws Exception {
        return invoke(method, params, requestTimeoutMs);
    }

    public Map<String, Object> invoke(String method, Map<String, Object> params, long timeoutMs) throws Exception {
        Objects.requireNonNull(method, "method");
        if (method.isBlank()) {
            throw new IllegalArgumentException("method is required");
        }

        awaitConnected(connectTimeoutMs);
        String id = "c" + nextID.incrementAndGet();
        CompletableFuture<Map<String, Object>> responseFuture = new CompletableFuture<>();
        pending.put(id, responseFuture);

        JsonObject payload = new JsonObject();
        payload.addProperty("jsonrpc", "2.0");
        payload.addProperty("id", id);
        payload.addProperty("method", method);
        payload.add("params", gson.toJsonTree(params != null ? params : Map.of()));

        try {
            send(payload);
        } catch (RuntimeException error) {
            pending.remove(id);
            throw error;
        }

        try {
            return responseFuture.get(timeoutMs, TimeUnit.MILLISECONDS);
        } finally {
            pending.remove(id);
        }
    }

    public synchronized void close() {
        closed = true;
        cancelHeartbeat();
        cancelReconnect();
        failAllPending(new IllegalStateException("holon-rpc client closed"));

        WebSocket ws = socket;
        socket = null;
        if (ws != null) {
            ws.sendClose(1000, "client close");
        }
    }

    private void awaitConnected(long timeoutMs) throws Exception {
        CompletableFuture<Void> cf = connectedFuture;
        try {
            cf.get(timeoutMs, TimeUnit.MILLISECONDS);
        } catch (ExecutionException ee) {
            Throwable cause = ee.getCause();
            if (cause instanceof Exception ex) {
                throw ex;
            }
            throw new RuntimeException(cause);
        }
    }

    private void openSocket(boolean initial) {
        if (closed) {
            return;
        }

        URI uri = URI.create(endpoint);
        httpClient.newWebSocketBuilder()
                .subprotocols("holon-rpc")
                .connectTimeout(Duration.ofMillis(connectTimeoutMs))
                .buildAsync(uri, new ListenerImpl())
                .whenComplete((ws, error) -> {
                    if (error != null) {
                        if (initial && !connectedFuture.isDone()) {
                            connectedFuture.completeExceptionally(error);
                        }
                        if (!closed) {
                            scheduleReconnect();
                        }
                    }
                });
    }

    private void handleIncoming(String text) {
        JsonElement parsed;
        try {
            parsed = JsonParser.parseString(text);
        } catch (Exception ignored) {
            return;
        }

        if (!parsed.isJsonObject()) {
            return;
        }

        JsonObject msg = parsed.getAsJsonObject();
        if (msg.has("method")) {
            handleRequest(msg);
            return;
        }
        if (msg.has("result") || msg.has("error")) {
            handleResponse(msg);
        }
    }

    private void handleRequest(JsonObject msg) {
        JsonElement id = msg.get("id");
        String jsonrpc = asString(msg.get("jsonrpc"));
        String method = asString(msg.get("method"));

        if (!"2.0".equals(jsonrpc) || method == null || method.isBlank()) {
            if (id != null && !id.isJsonNull()) {
                sendError(id, -32600, "invalid request", null);
            }
            return;
        }

        if ("rpc.heartbeat".equals(method)) {
            if (id != null && !id.isJsonNull()) {
                sendResult(id, Map.of());
            }
            return;
        }

        if (id != null && !id.isJsonNull()) {
            String sid = asID(id);
            if (sid == null || !sid.startsWith("s")) {
                sendError(id, -32600, "server request id must start with 's'", null);
                return;
            }
        }

        Handler handler = handlers.get(method);
        if (handler == null) {
            if (id != null && !id.isJsonNull()) {
                sendError(id, -32601, "method \"" + method + "\" not found", null);
            }
            return;
        }

        Map<String, Object> params = toMap(msg.get("params"));
        CompletionStage<Map<String, Object>> stage;
        try {
            stage = handler.handle(params);
        } catch (Exception error) {
            stage = CompletableFuture.failedFuture(error);
        }

        if (id == null || id.isJsonNull()) {
            return;
        }

        stage.whenComplete((result, error) -> {
            if (error != null) {
                Throwable cause = unwrap(error);
                if (cause instanceof HolonRPCResponseException rpcError) {
                    sendError(id, rpcError.code(), rpcError.getMessage(), rpcError.data());
                } else {
                    sendError(id, 13, cause.getMessage() != null ? cause.getMessage() : "internal error", null);
                }
                return;
            }
            sendResult(id, result != null ? result : Map.of());
        });
    }

    private void handleResponse(JsonObject msg) {
        String id = asID(msg.get("id"));
        if (id == null) {
            return;
        }

        CompletableFuture<Map<String, Object>> cf = pending.remove(id);
        if (cf == null || cf.isDone()) {
            return;
        }

        if (msg.has("error") && msg.get("error").isJsonObject()) {
            JsonObject errorObj = msg.getAsJsonObject("error");
            int code = errorObj.has("code") ? errorObj.get("code").getAsInt() : -32603;
            String message = errorObj.has("message") ? asString(errorObj.get("message")) : "internal error";
            Object data = errorObj.has("data") ? gson.fromJson(errorObj.get("data"), Object.class) : null;
            cf.completeExceptionally(new HolonRPCResponseException(code, message, data));
            return;
        }

        cf.complete(toMap(msg.get("result")));
    }

    private synchronized void handleDisconnect() {
        socket = null;
        cancelHeartbeat();
        failAllPending(new IllegalStateException("holon-rpc connection closed"));
        connectedFuture = new CompletableFuture<>();

        if (!closed) {
            scheduleReconnect();
        }
    }

    private synchronized void scheduleReconnect() {
        if (closed || reconnectTask != null) {
            return;
        }

        double base = Math.min(
                reconnectMinDelayMs * Math.pow(reconnectFactor, reconnectAttempt),
                reconnectMaxDelayMs);
        double jitter = base * reconnectJitter * ThreadLocalRandom.current().nextDouble();
        long delayMs = (long) (base + jitter);
        reconnectAttempt++;

        reconnectTask = scheduler.schedule(() -> {
            reconnectTask = null;
            openSocket(false);
        }, delayMs, TimeUnit.MILLISECONDS);
    }

    private synchronized void startHeartbeat() {
        cancelHeartbeat();
        heartbeatTask = scheduler.scheduleAtFixedRate(() -> {
            if (closed || socket == null) {
                return;
            }

            try {
                invoke("rpc.heartbeat", Map.of(), heartbeatTimeoutMs);
            } catch (Exception ignored) {
                WebSocket ws = socket;
                if (ws != null) {
                    ws.abort();
                }
            }
        }, heartbeatIntervalMs, heartbeatIntervalMs, TimeUnit.MILLISECONDS);
    }

    private synchronized void cancelHeartbeat() {
        if (heartbeatTask != null) {
            heartbeatTask.cancel(true);
            heartbeatTask = null;
        }
    }

    private synchronized void cancelReconnect() {
        if (reconnectTask != null) {
            reconnectTask.cancel(true);
            reconnectTask = null;
        }
        reconnectAttempt = 0;
    }

    private void send(JsonObject payload) {
        WebSocket ws = socket;
        if (ws == null) {
            throw new IllegalStateException("websocket is not connected");
        }
        ws.sendText(gson.toJson(payload), true).join();
    }

    private void sendResult(JsonElement id, Map<String, Object> result) {
        JsonObject payload = new JsonObject();
        payload.addProperty("jsonrpc", "2.0");
        payload.add("id", id);
        payload.add("result", gson.toJsonTree(result != null ? result : Map.of()));
        send(payload);
    }

    private void sendError(JsonElement id, int code, String message, Object data) {
        JsonObject payload = new JsonObject();
        payload.addProperty("jsonrpc", "2.0");
        payload.add("id", id);
        JsonObject error = new JsonObject();
        error.addProperty("code", code);
        error.addProperty("message", message);
        if (data != null) {
            error.add("data", gson.toJsonTree(data));
        }
        payload.add("error", error);
        send(payload);
    }

    private void failAllPending(Throwable error) {
        if (pending.isEmpty()) {
            return;
        }
        Map<String, CompletableFuture<Map<String, Object>>> snapshot = Map.copyOf(pending);
        pending.clear();
        snapshot.values().forEach(cf -> cf.completeExceptionally(error));
    }

    private Map<String, Object> toMap(JsonElement element) {
        if (element == null || element.isJsonNull()) {
            return Map.of();
        }
        if (!element.isJsonObject()) {
            return Map.of();
        }
        Map<String, Object> result = gson.fromJson(element, MAP_TYPE);
        return result != null ? result : Map.of();
    }

    private static String asString(JsonElement element) {
        if (element == null || element.isJsonNull()) {
            return null;
        }
        return element.getAsString();
    }

    private static String asID(JsonElement element) {
        if (element == null || element.isJsonNull()) {
            return null;
        }
        if (element.isJsonPrimitive()) {
            return element.getAsString();
        }
        return null;
    }

    private static Throwable unwrap(Throwable throwable) {
        if (throwable instanceof CompletionException ce && ce.getCause() != null) {
            return ce.getCause();
        }
        return throwable;
    }

    private final class ListenerImpl implements WebSocket.Listener {
        private final StringBuilder textBuffer = new StringBuilder();

        @Override
        public void onOpen(WebSocket webSocket) {
            String protocol = webSocket.getSubprotocol();
            if (!"holon-rpc".equals(protocol)) {
                if (!connectedFuture.isDone()) {
                    connectedFuture.completeExceptionally(
                            new IllegalStateException("server did not negotiate holon-rpc subprotocol"));
                }
                webSocket.sendClose(1002, "missing holon-rpc subprotocol");
                return;
            }

            socket = webSocket;
            reconnectAttempt = 0;
            cancelReconnect();
            startHeartbeat();
            if (!connectedFuture.isDone()) {
                connectedFuture.complete(null);
            }
            webSocket.request(1);
        }

        @Override
        public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
            textBuffer.append(data);
            if (last) {
                handleIncoming(textBuffer.toString());
                textBuffer.setLength(0);
            }
            webSocket.request(1);
            return CompletableFuture.completedFuture(null);
        }

        @Override
        public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
            handleDisconnect();
            return CompletableFuture.completedFuture(null);
        }

        @Override
        public void onError(WebSocket webSocket, Throwable error) {
            handleDisconnect();
        }
    }
}

File: sdk/kotlin-holons/src/main/kotlin/org/organicprogramming/holons/HolonRPCClient.kt
package org.organicprogramming.holons

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import kotlin.math.min
import kotlin.random.Random

class HolonRPCResponseException(
    val code: Int,
    override val message: String,
    val data: JsonElement? = null,
) : RuntimeException("rpc error $code: $message")

class HolonRPCClient(
    private val heartbeatIntervalMs: Long = 15_000,
    private val heartbeatTimeoutMs: Long = 5_000,
    private val reconnectMinDelayMs: Long = 500,
    private val reconnectMaxDelayMs: Long = 30_000,
    private val reconnectFactor: Double = 2.0,
    private val reconnectJitter: Double = 0.1,
    private val connectTimeoutMs: Long = 10_000,
    private val okHttpClient: OkHttpClient = OkHttpClient.Builder().readTimeout(0, java.util.concurrent.TimeUnit.MILLISECONDS).build(),
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) {
    private val json = Json { ignoreUnknownKeys = false }

    private val handlers = ConcurrentHashMap<String, suspend (JsonObject) -> JsonObject>()
    private val pending = ConcurrentHashMap<String, CompletableDeferred<JsonObject>>()
    private val nextID = AtomicLong(0)

    private val connectionState = MutableStateFlow(false)
    private val reconnecting = AtomicBoolean(false)
    private val closed = AtomicBoolean(false)

    @Volatile
    private var endpoint: String? = null

    @Volatile
    private var webSocket: WebSocket? = null

    @Volatile
    private var heartbeatJob: Job? = null

    @Volatile
    private var reconnectJob: Job? = null

    suspend fun connect(url: String) {
        endpoint = url
        closed.set(false)
        openWebSocket(url)
        awaitConnected(connectTimeoutMs)
    }

    fun register(method: String, handler: suspend (JsonObject) -> JsonObject) {
        require(method.isNotBlank()) { "method is required" }
        handlers[method] = handler
    }

    suspend fun invoke(
        method: String,
        params: JsonObject = buildJsonObject { },
        timeoutMs: Long = 10_000,
    ): JsonObject {
        require(method.isNotBlank()) { "method is required" }
        awaitConnected(connectTimeoutMs)

        val id = "c${nextID.incrementAndGet()}"
        val deferred = CompletableDeferred<JsonObject>()
        pending[id] = deferred

        val payload = buildJsonObject {
            put("jsonrpc", JsonPrimitive("2.0"))
            put("id", JsonPrimitive(id))
            put("method", JsonPrimitive(method))
            put("params", params)
        }

        val sent = webSocket?.send(payload.toString()) ?: false
        if (!sent) {
            pending.remove(id)
            throw IllegalStateException("websocket send failed")
        }

        return try {
            withTimeout(timeoutMs) {
                deferred.await()
            }
        } finally {
            pending.remove(id)
        }
    }

    suspend fun close() {
        closed.set(true)
        reconnectJob?.cancel()
        reconnectJob = null

        heartbeatJob?.cancel()
        heartbeatJob = null

        webSocket?.close(1000, "client close")
        webSocket = null
        connectionState.value = false

        failAllPending(IllegalStateException("holon-rpc client closed"))
        scope.cancel()

        okHttpClient.dispatcher.cancelAll()
        okHttpClient.dispatcher.executorService.shutdownNow()
        okHttpClient.connectionPool.evictAll()
        okHttpClient.cache?.close()
    }

    private suspend fun awaitConnected(timeoutMs: Long) {
        if (connectionState.value) {
            return
        }

        withTimeout(timeoutMs) {
            connectionState.filter { it }.first()
        }
    }

    private fun openWebSocket(url: String) {
        val request = Request.Builder()
            .url(url)
            .header("Sec-WebSocket-Protocol", "holon-rpc")
            .build()

        okHttpClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                val protocol = response.header("Sec-WebSocket-Protocol")
                if (protocol != "holon-rpc") {
                    webSocket.close(1002, "missing holon-rpc subprotocol")
                    return
                }

                this@HolonRPCClient.webSocket = webSocket
                connectionState.value = true
                startHeartbeat()
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                scope.launch {
                    handleIncomingMessage(text)
                }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(code, reason)
                scope.launch { handleDisconnect() }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                scope.launch { handleDisconnect() }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                scope.launch { handleDisconnect() }
            }
        })
    }

    private suspend fun handleIncomingMessage(text: String) {
        val payload = json.parseToJsonElement(text).jsonObject

        if (payload["method"] != null) {
            handleIncomingRequest(payload)
        } else if (payload["result"] != null || payload["error"] != null) {
            handleIncomingResponse(payload)
        }
    }

    private suspend fun handleIncomingRequest(payload: JsonObject) {
        val method = payload["method"]?.jsonPrimitive?.content
        val id = payload["id"]
        val jsonrpc = payload["jsonrpc"]?.jsonPrimitive?.content

        if (jsonrpc != "2.0" || method == null) {
            if (id != null) {
                sendError(id, -32600, "invalid request")
            }
            return
        }

        if (method == "rpc.heartbeat") {
            if (id != null) {
                sendResult(id, buildJsonObject { })
            }
            return
        }

        if (id != null) {
            val sid = id.jsonPrimitive.content
            if (!sid.startsWith("s")) {
                sendError(id, -32600, "server request id must start with 's'")
                return
            }
        }

        val handler = handlers[method]
        if (handler == null) {
            if (id != null) {
                sendError(id, -32601, "method \"$method\" not found")
            }
            return
        }

        val params = payload["params"]?.jsonObject ?: buildJsonObject { }

        try {
            val result = handler(params)
            if (id != null) {
                sendResult(id, result)
            }
        } catch (rpc: HolonRPCResponseException) {
            if (id != null) {
                sendError(id, rpc.code, rpc.message, rpc.data)
            }
        } catch (t: Throwable) {
            if (id != null) {
                sendError(id, 13, t.message ?: "internal error")
            }
        }
    }

    private fun handleIncomingResponse(payload: JsonObject) {
        val id = payload["id"]?.jsonPrimitive?.content ?: return
        val deferred = pending.remove(id) ?: return

        val errorObj = payload["error"]?.jsonObject
        if (errorObj != null) {
            val code = errorObj["code"]?.jsonPrimitive?.int ?: -32603
            val message = errorObj["message"]?.jsonPrimitive?.content ?: "internal error"
            deferred.completeExceptionally(HolonRPCResponseException(code, message, errorObj["data"]))
            return
        }

        val result = payload["result"]?.jsonObject ?: buildJsonObject { }
        deferred.complete(result)
    }

    private fun sendResult(id: JsonElement, result: JsonObject) {
        val payload = buildJsonObject {
            put("jsonrpc", JsonPrimitive("2.0"))
            put("id", id)
            put("result", result)
        }
        if (webSocket?.send(payload.toString()) != true) {
            throw IllegalStateException("websocket send failed")
        }
    }

    private fun sendError(id: JsonElement, code: Int, message: String, data: JsonElement? = null) {
        val payload = buildJsonObject {
            put("jsonrpc", JsonPrimitive("2.0"))
            put("id", id)
            put("error", buildJsonObject {
                put("code", JsonPrimitive(code))
                put("message", JsonPrimitive(message))
                if (data != null) {
                    put("data", data)
                }
            })
        }
        if (webSocket?.send(payload.toString()) != true) {
            throw IllegalStateException("websocket send failed")
        }
    }

    private fun startHeartbeat() {
        if (heartbeatJob?.isActive == true) {
            return
        }

        heartbeatJob = scope.launch {
            while (true) {
                delay(heartbeatIntervalMs)
                if (closed.get()) {
                    return@launch
                }

                try {
                    withTimeout(heartbeatTimeoutMs) {
                        invoke("rpc.heartbeat", buildJsonObject { })
                    }
                } catch (_: Throwable) {
                    webSocket?.cancel()
                    return@launch
                }
            }
        }
    }

    private suspend fun handleDisconnect() {
        webSocket = null
        connectionState.value = false
        failAllPending(IllegalStateException("holon-rpc connection closed"))

        if (!closed.get()) {
            startReconnectLoop()
        }
    }

    private fun startReconnectLoop() {
        if (!reconnecting.compareAndSet(false, true)) {
            return
        }

        reconnectJob?.cancel()
        reconnectJob = scope.launch {
            var attempt = 0

            while (!closed.get()) {
                val target = endpoint ?: break
                openWebSocket(target)

                try {
                    awaitConnected(connectTimeoutMs)
                    reconnecting.set(false)
                    return@launch
                } catch (_: Throwable) {
                    val base = min(
                        reconnectMinDelayMs * Math.pow(reconnectFactor, attempt.toDouble()),
                        reconnectMaxDelayMs.toDouble(),
                    )
                    val jitter = base * reconnectJitter * Random.nextDouble(0.0, 1.0)
                    delay((base + jitter).toLong())
                    attempt += 1
                }
            }

            reconnecting.set(false)
        }
    }

    private fun failAllPending(error: Throwable) {
        val values = pending.values.toList()
        pending.clear()
        values.forEach { it.completeExceptionally(error) }
    }
}
